kube-runtime/src/controller/runner.rs

```diff
@@ -0,0 +1,128 @@
+use super::stream_hash_map::StreamHashMap;
+use crate::scheduler::{self, ScheduleRequest, Scheduler};
+use futures::{Future, Stream, StreamExt};
+use pin_project::pin_project;
+use std::{
+    hash::Hash,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// Pulls messages from a [`Scheduler`], and runs an action for each message in parallel,
+/// while making sure to not run the same message multiple times at once.
+#[pin_project]
+pub struct Runner<T, R, F, MkF> {
```

Comment:
> would mean these two changes will be executed in parallel, unless the implementor of the reconciliation loop (aka user of kube-rs library) takes care of that and implements it's own queue.

Agreed, avoiding that case was my Unbreakable Rule for doing this.

The `Runner` (and, in turn, the `Scheduler`) only ever sees the `ObjectRef` (`name`+`namespace`+`group-version-kind`). That should be the same for both of your cases, ensuring that they are serialized.

Reply:
We could make the `Runner` pluggable so that they could plug in their own logic (to avoid the `StreamHashMap` overhead for the 100% sequential use-case, for example).

For limited parallelism, I would say you should just put a finer-grained `Semaphore` in front of the limited resource you're trying to access. I don't think building that into the `Runner` itself would be a good tradeoff, we would still need the existing bookkeeping, and we wouldn't be able to do much with the extra information (for example, we can't dedupe events between different objects anyway..).

Reply:
Great. Well, with ObjectRef compared it makes sense. Thank you.

Reply:
Yeah, it would probably be hard to put in the `Runner`. I can't think of a good solution to it as is, but I do feel it's probably something we should consider having a solution to in the future, as this stream is pretty opaque from a users POV. Still, it's not worth blocking this over.